cmd/jujud/agent/caasoperator_test.go

```diff
@@ -0,0 +1,228 @@
+		{"--application-name", "wordpress/wild/9"},
+		{"--application-name", "20"},
+	} {
+		a, err := NewCaasOperatorAgent(nil, s.newBufferedLogWriter())
+		c.Assert(err, jc.ErrorIsNil)
+
+		err = cmdtesting.InitCommand(a, append(args, "--data-dir", "jc"))
+		c.Check(err, gc.ErrorMatches, `--application-name option expects "<application>" argument`)
+	}
+}
+
+func (s *CAASOperatorSuite) TestParseUnknown(c *gc.C) {
+	a, err := NewCaasOperatorAgent(nil, s.newBufferedLogWriter())
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = cmdtesting.InitCommand(a, []string{
+		"--application-name", "wordpress",
+		"thundering typhoons",
+	})
+	c.Check(err, gc.ErrorMatches, `unrecognized args: \["thundering typhoons"\]`)
+}
+
+func waitForApplicationActive(c *gc.C, agentDir string) {
+	timeout := time.After(coretesting.LongWait)
+
+	for {
+		select {
+		case <-timeout:
+			c.Fatalf("no activity detected")
+		case <-time.After(coretesting.ShortWait):
+			link := filepath.Join(agentDir, commands.CommandNames()[0])
+			if _, err := os.Lstat(link); err == nil {
+				target, err := os.Readlink(link)
+				c.Assert(err, jc.ErrorIsNil)
+				c.Assert(target, gc.Equals, filepath.Join(filepath.Dir(os.Args[0]), "jujud"))
+				return
+			}
+		}
+	}
+}
+
+func (s *CAASOperatorSuite) TestRunStop(c *gc.C) {
+	app, config, _ := s.primeAgent(c)
+	a := s.newAgent(c, app)
+	go func() { c.Check(a.Run(nil), gc.IsNil) }()
+	defer func() { c.Check(a.Stop(), gc.IsNil) }()
+	waitForApplicationActive(c, filepath.Join(config.DataDir(), "tools"))
+}
+
+func (s *CAASOperatorSuite) TestOpenStateFails(c *gc.C) {
+	app, config, _ := s.primeAgent(c)
+	a := s.newAgent(c, app)
+	go func() { c.Check(a.Run(nil), gc.IsNil) }()
+	defer func() { c.Check(a.Stop(), gc.IsNil) }()
+	waitForApplicationActive(c, filepath.Join(config.DataDir(), "tools"))
+
+	s.AssertCannotOpenState(c, config.Tag(), config.DataDir())
+}
+
+func (s *CAASOperatorSuite) TestUseLumberjack(c *gc.C) {
+	ctx, err := cmd.DefaultContext()
+	c.Assert(err, gc.IsNil)
+
+	a := CaasOperatorAgent{
+		AgentConf:       FakeAgentConfig{},
+		ctx:             ctx,
+		ApplicationName: "mysql",
+	}
+
+	err = a.Init(nil)
+	c.Assert(err, gc.IsNil)
+
+	l, ok := ctx.Stderr.(*lumberjack.Logger)
+	c.Assert(ok, jc.IsTrue)
+	c.Check(l.MaxAge, gc.Equals, 0)
+	c.Check(l.MaxBackups, gc.Equals, 2)
+	c.Check(l.Filename, gc.Equals, filepath.FromSlash("/var/log/juju/machine-42.log"))
+	c.Check(l.MaxSize, gc.Equals, 300)
+}
+
+func (s *CAASOperatorSuite) TestDontUseLumberjack(c *gc.C) {
+	ctx, err := cmd.DefaultContext()
+	c.Assert(err, gc.IsNil)
+
+	a := CaasOperatorAgent{
+		AgentConf:       FakeAgentConfig{},
+		ctx:             ctx,
+		ApplicationName: "mysql",
+
+		// this is what would get set by the CLI flags to tell us not to log to
+		// the file.
+		logToStdErr: true,
+	}
+
+	err = a.Init(nil)
+	c.Assert(err, gc.IsNil)
+
+	_, ok := ctx.Stderr.(*lumberjack.Logger)
+	c.Assert(ok, jc.IsFalse)
+}
+
+func (s *CAASOperatorSuite) TestWorkers(c *gc.C) {
```

Comment:
I think we should replace this with a combination of manifold name tests (i.e. in the manifold package), and feature tests.

Reply:
The unit tests have been left in place and this moved to a feature test. It tests the machinery that starts the workers so best to retain it (as well as existing name tests).